### PR TITLE
feat(graph): support native Cargo projects

### PR DESCRIPTION
**Pull request by pepicrft**
## What changed

Once now recognizes a root `Cargo.toml` as a native project and derives a typed build graph without requiring `once.toml`.

- Added native Cargo workspace discovery, locked metadata resolution, workspace member targets, external package targets, build scripts, feature selection, binaries, libraries, and tests.
- Added generic host-tree materialization and race-safe restoration of cached directory outputs.
- Added compact test result summaries to the command-line and [Model Context Protocol](https://modelcontextprotocol.io/) surfaces so agents can read large test suites without receiving every case.
- Documented the incremental path from Cargo-only discovery to a recorded native seed, explicit typed targets, and reusable Starlark modules.

## Why

Rust projects should be able to adopt Once with their existing Cargo manifests. A repository can begin with no Once configuration, gain cached target-level builds and tests, and add explicit graph ownership only where it needs more control.

The implementation also needs to preserve Once's ecosystem-neutral architecture. Cargo-specific discovery and execution behavior belongs in Starlark target kinds, while the Rust implementation provides generic graph, action, materialization, validation, and cache primitives.

## Root cause

Native project discovery previously covered Mix but did not expose Cargo as a built-in native project. Cargo dependency sources also needed a portable way to enter the action graph without assuming a repository-managed vendor directory.

Kura's parallel tests exposed a separate cached-output race: concurrent actions could restore the same directory output simultaneously. Large normalized test results also exceeded agent response limits because the query always returned every case.

## Approach

The Cargo workspace target kind reads locked Cargo metadata and expands it into the existing Rust target kinds. External source trees are content-verified during analysis and execution through a generic host-tree action. The shared native-project loader remains unaware of Cargo, Mix, Rust, or Elixir.

Cached output restoration now validates directory digests, serializes the materialization section per workspace, and validates again after taking the lock. Compact test summaries preserve status, totals, runner information, and artifacts while omitting case arrays.

## Impact

Users can run Once directly from a Cargo workspace, query the generated targets, build binaries, and run normalized tests without creating `once.toml`. Existing hand-authored Rust targets and recorded Cargo seeds remain supported.

Projects can adopt configuration incrementally. Agent workflows can retrieve exact totals from large test suites without truncation.

## Validation

- `mise exec -- cargo test --workspace`
- `mise exec -- cargo test -p once-cli`
- `mise exec -- cargo clippy --workspace --all-targets -- -D warnings`
- `mise exec -- cargo fmt --all -- --check`
- Documentation tests: 6 passed
- Elixir web tests through Once: 52 passed
- Native Mix project without `once.toml`: passed
- Kura native Cargo build: passed
- Kura library tests: 343 passed, 1 skipped, 0 failed
- Kura binary tests: passed with 0 discovered cases
- Claude, restricted to Once's Model Context Protocol tools, detected the root `Cargo.toml`, confirmed that `once.toml` was absent, built Kura, ran both test targets, and retrieved complete compact summaries without truncation
- Local Rust documentation route returned HTTP status 200

### CODE REVIEW

File: crates/once-core/src/execute.rs [hunk 1]
```diff
@@ -1003,6 +1062,65 @@ fn file_sha256_hex(path: &Path) -> std::io::Result<String> {
...
+            b'f'
+        } else {
+            continue;
+        };
+        hasher.update([kind]);
+        hasher.update(relative.as_bytes());
+        hasher.update([0]);
+        hasher.update(host_file_mode(&metadata).to_le_bytes());
+        if metadata.file_type().is_symlink() {
+            hasher.update(std::fs::read_link(&path)?.to_string_lossy().as_bytes());
```
[the remark this answers is not in the document]
Comment (pepicrft) [reply crates/once-core/src/execute.rs]: Fixed in d2302a5. The tree digest now rejects any symlink whose target resolves outside the tree, because output capture (MaterializeExternal) materializes those external contents while the digest only hashed the link text, so external contents could change without moving the action key. Links that stay inside the tree are unchanged: they hash to link text on both the input digest and MaterializeExternal capture, so they were already consistent. Dangling links stay accepted since they carry no external contents to miss. The hashing lives in one shared crate, so analysis and execution reject identically. Added tests for internal, external, and dangling links.

File: crates/once-core/src/execute.rs [hunk 2]
```diff
@@ -635,6 +645,55 @@ async fn materialize_host_file(
...
+            source: source_error,
+        })?;
+        if actual != expected {
+            return Err(Error::HostFileDigestMismatch {
+                path: source.display().to_string(),
+                expected,
+                actual,
+            });
+        }
+        copy_sandbox_output_blocking(&source, &destination_path).map_err(|source_error| {
```
[the remark this answers is not in the document]
Comment (pepicrft) [reply crates/once-core/src/execute.rs:688]: Fixed in 60fff8d. materialize_host_tree now rejects an overlapping source and destination before touching disk: equal paths, a destination nested under the source, or a source nested under the destination. The check canonicalizes the existing prefixes of both paths so it cannot be sidestepped by linking one side into the other. Added a test asserting that a destination nested under the source is rejected and the source data is left intact.

[the remark this answers is not in the document]
Comment (pepicrft) [reply crates/once-core/src/execute.rs:688]: Fixed in 60fff8d. After copying the source into the destination we re-hash the copied snapshot and reject it when it no longer matches the expected digest, so a source that changed between the initial hash and the copy can no longer be captured under a stale key. std::fs::copy preserves permission bits and symlinks are recreated faithfully, so an unchanged source re-hashes to the same digest and the happy path is unaffected (verified by the existing replay-from-cache test).

File: web/priv/docs/guide/graph/rust.md [hunk 3]
```diff
@@ -5,63 +5,224 @@ next: false
 
 # Rust
 
-Once can build Rust libraries, binaries, tests, procedural macros, and native
-mobile libraries. This guide starts with one local library, a binary that uses
-it, and a test for the same code. Cargo dependency resolution comes after that
-first working graph.
+Once can read an existing `Cargo.toml`, derive a typed build graph, and cache
+each workspace package and locked dependency separately. You can build, run,
+and test the project without first translating it into `once.toml`.
```
[the remark this answers is not in the document]
Comment (pepicrft) [reply web/priv/docs/guide/graph/rust.md:10]: Added web/priv/changelog/2026-07-29-native-cargo-project-graphs.md in 7741cab, following the existing native Mix entry, so native Cargo support appears on /changelog and in the RSS/Atom feeds.

File: crates/once-cli/src/commands/mcp/tools.rs [hunk 4]
```diff
@@ -414,6 +414,10 @@ pub fn tool_catalog() -> Vec<ToolDefinition> {
                     "test_unit": {
                         "type": "string",
                         "description": "Run one stable unit identifier returned by once_query_test_manifest. Requires exactly one explicit target."
+                    },
+                    "summary_only": {
+                        "type": "boolean",
+                        "description": "Replace case-level normalized results with compact once.test_results_summary.v1 totals."
```
[the remark this answers is not in the document]
Comment (pepicrft) [reply crates/once-cli/src/commands/mcp/tools.rs:420]: Not changing this. The convention is that every MCP tool has a matching once query or once edit subcommand so a human can reproduce it. The tool that carries summary_only over completed results, once_query_test_results, already has its CLI twin: once query test-results --summary-only (the summary_only flag exists on the TestResults query command in cli/query.rs). So the compact projection is reproducible from the terminal today: run once test <target>, then once query test-results <target> --summary-only.

once_run_tests bundles the run and the report into a single call for the agent, and its summary_only compacts the report's per-run results object. once test renders through a different progress-and-results path (test_schedule::run) and emits no such report object for a flag to project. Since the projection is already available on the query command, adding a second surface on once test would duplicate it rather than close a real parity gap.

File: crates/once-cli/src/cli/query.rs [hunk 5]
```diff
@@ -116,6 +116,9 @@ pub enum QueryCmd {
     TestResults {
         /// Target id, such as `tests/unit`.
         target: String,
+        /// Return the normalized status and totals without case-level records.
+        #[arg(long)]
+        summary_only: bool,
```
Comment (pepicrft) [inline crates/once-cli/src/cli/query.rs:121]: The new `--summary-only` projection has unit coverage for `test_results_summary_value`, but no ShellSpec exercises the flag end to end. The repo's own review rule asks for ShellSpec coverage on user-visible CLI changes, and the existing `queries normalized shellspec test results` block in `spec/test_capability_spec.sh` is the natural neighbor. A small spec asserting that `once query test-results <target> --summary-only` returns the `once.test_results_summary.v1` schema and omits case records would close the gap. Non-blocking; flagging for consistency with the other CLI surfaces.

**Review by pepicrft — COMMENTED** [review]
Reviewed this independently and the safety-critical paths hold up.

Verified the cached-output race fix. `restore` now takes a workspace-global `fd_lock` write lock around the materialization loop and re-checks each output digest inside the lock. I confirmed that `fd_lock` (flock-based) actually denies a second exclusive lock taken from a separate open within the same process, so concurrent restores inside one process serialize correctly, which is the case that matters for the scheduler.

The host-tree work is sound. `source_sha256` participates in the action digest so cache identity is content-addressed, external symlinks are rejected in the tree digest, the overlap check guards against copying a tree into itself, and the post-copy re-hash closes the time-of-check/time-of-use window that Codex flagged. Each of those has a focused test in `once-core`.

`parse_statuses` is a real fix rather than a refactor. libtest emits `test x ... ignored, <reason>` for `#[ignore = "..."]`, and the old suffix match missed those cases; the new `split_once` with the empty or `, ` check matches both forms, which I confirmed against real libtest output.

clippy, fmt, the once-core, once-frontend, and once-cli test subsets, the examples integration test, and the new host-tree crate tests are all green. Changelog, docs route, and release manifest wiring are complete.

One non-blocking note inline on `--summary-only` missing ShellSpec coverage. Approving; that spec can land in a follow-up.

File: crates/once-cli/src/cli/query.rs (hunk 5 above)
Comment (pepicrft) [inline crates/once-cli/src/cli/query.rs:121]: The new `--summary-only` projection has unit coverage for `test_results_summary_value`, but no ShellSpec exercises the flag end to end. The repo's own review rule asks for ShellSpec coverage on user-visible CLI changes, and the existing `queries normalized shellspec test results` block in `spec/test_capability_spec.sh` is the natural neighbor. A small spec asserting that `once query test-results <target> --summary-only` returns the `once.test_results_summary.v1` schema and omits case records would close the gap. Non-blocking; flagging for consistency with the other CLI surfaces.

Comment (pepicrft) [reply crates/once-cli/src/cli/query.rs:121]: Addressed in d2d2552. Added a ShellSpec next to the existing normalized-results query that runs `once query test-results cli_e2e --summary-only`, asserts the `once.test_results_summary.v1` schema with the `once.test_results.v1` source schema, target, and status, and confirms case-level records (the `spec/cli_surface_spec.sh::lists commands` case the full query returns) are omitted. Verified it passes against a release build.
